Mathlib/AlgebraicTopology/SimplicialSet/HomotopyCat.lean

```diff
@@ -0,0 +1,469 @@
+
+The homotopy category of a simplicial set is defined as a quotient of the free category on its
+underlying reflexive quiver (equivalently its one truncation). The quotient imposes an additional
+hom relation on this free category, asserting that `f ≫ g = h` whenever `f`, `g`, and `h` are
+respectively the 2nd, 0th, and 1st faces of a 2-simplex.
+
+In this file, we in fact define a pair of functors:
+
+(1) `SSet.hoFunctor' : SSet.{u} ⥤ Cat.{u, u}` implements the construction described above, while
+
+(2) `SSet.hoFunctor : SSet.{u} ⥤ Cat.{u, u} := SSet.truncation 2 ⋙ SSet.hoFunctor₂` is defined by
+first restricting from simplicial sets to 2-truncated simplicial sets (throwing away the data that
+is not used for the construction of the homotopy category) and then composing with an analogously
+defined `SSet.hoFunctor₂ : SSet.Truncated.{u} 2 ⥤ Cat.{u,u}` implemented relative to the syntax of
+the 2-truncated simplex category.
+
+It should be relatively straightforward to show that these constructions agree:
+
+def hoFunctor.ofTwoTruncation.iso (V : SSet) :
+    SSet.hoFunctor₂Obj ((truncation 2).obj V) ≅ SSet.hoCat V := sorry
+
+def hoFunctor.ofTwoTruncation.natIso :
+     truncation 2 ⋙ SSet.hoFunctor₂ ≅ SSet.hoFunctor' := sorry
+
+but we leave this for future work.
+
+The functor `SSet.hoFunctor` is shown to be left adjoint to the nerve by providing an analogous
+decomposition of the nerve functor and then composing a pair of adjunctions, which factor through
+the category of 2-truncated simplicial sets.
+-/
+
+namespace CategoryTheory
+open Category Limits Functor Opposite Simplicial
+universe v u
+
+section
+
+local notation3:1000 (priority := high) X " _[" n "]" =>
+    (X : CategoryTheory.SimplicialObject _).obj (Opposite.op (SimplexCategory.mk n))
+
+/-- A simplicial set `S` has an underlying refl quiver with `S _[0]` as its underlying type.-/
+def SSet.OneTruncation (S : SSet) := S _[0]
+
+/-- The source vertex of `f : S _[1]` for use in defining the underlying refl quiver.-/
+def SSet.OneTruncation.src {S : SSet} (f : S _[1]) : OneTruncation S :=
+  S.map (SimplexCategory.δ (n := 0) 1).op f
+
+/-- The target vertex of `f : S _[1]` for use in defining the underlying refl quiver.-/
+def SSet.OneTruncation.tgt {S : SSet} (f : S _[1]) : OneTruncation S :=
+  S.map (SimplexCategory.δ (n := 0) 0).op f
+
+/-- The hom-types of the refl quiver underlying a simplicial set `S` are subtypes of `S _[1]`.-/
+def SSet.OneTruncation.Hom {S : SSet} (X Y : OneTruncation S) :=
+  {p : S _[1] // src p = X ∧ tgt p = Y}
+
+/-- A simplicial set `S` has an underlying refl quiver `SSet.OneTruncation S`.-/
+instance (S : SSet) : ReflQuiver (SSet.OneTruncation S) where
+  Hom X Y := SSet.OneTruncation.Hom X Y
+  id X := by
+    refine ⟨S.map (SimplexCategory.σ (n := 0) 0).op X, ?_, ?_⟩ <;>
+    · change (S.map _ ≫ S.map _) X = X
+      rw [← map_comp, (_ : _ ≫ _ = 𝟙 _)]; simp
+      show ({..} : Opposite _) = _; congr; ext i
+      let 0 := i
+      rfl
```

Comment:
The API for simplicial sets should be expanded to include concrete formulations of the simplicial relations, like
```lean
@[simp]
lemma SSet.δ_comp_σ_self_apply (S : SSet) {n} (i : Fin (n + 1)) (x : S _[n]) :
    S.δ i.castSucc (S.σ i x) = x :=
  congr_fun S.δ_comp_σ_self x
```

Reply:
(This could go in a separate PR.)

Reply:
Done except for the two naturality statements, which I forgot about, and added here.